---
title: "Metagenome RPKM plots"
author: "Connor Morgan-Lang"
date: "November 26, 2018"
output: html_document
---

This bit of code will visualize the quality and relative abundance of your MAGs. We rely on the checkM output table for completeness and contamination quality metrics, the gtdbtk output for taxonomic identity, and the Metagenome RPKM file that was generated by the `bin_multi.sh` script (`binned.rpkm.csv`). After a bit of data frame joining we will be able to visualize these data using `ggplot`.

__NOTE:__ The plots that I make are not the "answers" or ideal plots we are looking for, they are just examples. These are probably not the best for answering your group's specific questions and so you may find yourself modifying axis limits, filtering by different quality thresholds and focusing on different taxa. This is great! Don't be afraid to explore a bit.

Before you begin loading, transforming, and playing with data we will need to load the following libraries at the very least:

```{r setup, warning=FALSE, message=FALSE}
library(tidyr)
library(dplyr)
library(RColorBrewer)
library(knitr)
library(ggplot2)
```

Now we are going to read our tabular files. These commands are also in the pathview Rmarkdown tutorial.

```{r read, warning=FALSE, message=FALSE}
arc_class <- read.table("~/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/testing/gtdbtk.ar122.classification_pplacer.tsv", sep="\t")
bac_class <- read.table("~/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/testing/gtdbtk.bac120.classification_pplacer.tsv", sep="\t")
gtdb_dat <- rbind(arc_class, bac_class) %>% 
  dplyr::rename(mag = V1) %>% 
  separate(V2, sep=';', into=c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>%
  dplyr::select(mag, Kingdom, Phylum, Class, Order, Family)

checkm_dat <- read.table("~/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/testing/MetaBAT2_SaanichInlet_150m_min1500_checkM_stdout.tsv",
                         header=TRUE,
                         sep="\t",
                         comment.char = '') %>% 
  dplyr::rename(mag = Bin.Id) %>% 
  dplyr::select(mag, Completeness, Contamination)

# Due to a bug in the renaming script we have to rename the bins. Its a bit hacky but works using tidyverse functions
metag_rpkm <- read.table("~/Desktop/MICB405_TAship/MICB405-Metagenomics/2018/testing/SaanichInlet_150m_binned.rpkm.csv", header=T, sep=',') %>% 
  mutate(Sequence = gsub('m_', 'm.', Sequence)) %>% 
  mutate(Sequence = gsub('Inlet_', 'Inlet.', Sequence)) %>% 
  separate(col=Sequence, into=c("mag", "contig"), sep='_', extra="merge") %>% 
  group_by(Sample, mag) %>% 
  summarise(g_rpkm = sum(RPKM)) %>% 
  mutate(mag = gsub('Inlet.', 'Inlet_', mag))

```

Here we join the metagenome RPKM, checkM, and GTDB-tk data frames, then take the mean RPKM of each MAG across the different cruises. You can perform this analysis and the `pathview` analysis using the same dataframe but we're just going to make ours specific for the quality and abundance bubble plots. 

```{r }
rpkm_dat <- left_join(metag_rpkm, checkm_dat, by="mag") %>% 
  left_join(gtdb_dat, by="mag") %>% 
  group_by(mag, Kingdom, Phylum, Class, Completeness, Contamination) %>% 
  summarise(g_rpkm = mean(g_rpkm))
```
Now onto plotting these data with `ggplot`. The x-axis is completeness, y-axis is contamination and the points are coloured by a MAG's taxonomic class (each bubble represents a MAG). Also, I've decided to scale the size of the points, or bubbles, by a MAG's mean RPKM value.
I've also decided to bound the completeness from 50-100% and the contamination from 0-100 (there are values >100).
The arguments in the `theme()` layer change the background to my favourite style. None of this gray background for me! More negative space, please.

```{r plot}
ggplot(rpkm_dat, aes(x=Completeness, y=Contamination, col=Class)) +
  geom_point(aes(size=g_rpkm)) +
  scale_size(range=c(1,10)) +
  xlim(c(50,100)) +
  ylim(c(0,100)) +
  theme(panel.background = element_blank(),
        panel.grid.major = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank())
```

For this next plot, we will determine the relative abundance of these MAGs across the different cruises and time points. We don't want to use the mean RPKM value across the cruises so we are going to create a new dataframe for this plot. We'll begin exactly as before, but we're going to filter the MAGs down to something that isn't crowded and visually overwhelming. 
I selected all the medium- and high-quality Proteobacteria, then sorted by taxonomic Order. This will make `ggplot` plot all the MAGs that are taxonomically related together, rather than by their arbitrary numerical identifiers. We're also discarding all the low-quality MAGs since they are full of sequences from who-knows-what. Total Frankenbacteria!


```{r cruise_bubbles}
rpkm_dat <- left_join(metag_rpkm, checkm_dat, by="mag") %>% 
  left_join(gtdb_dat, by="mag") %>% 
  filter(Completeness > 50 & Contamination < 10) %>% # good quality MAGs
  filter(Phylum == "p__Proteobacteria") %>% # Proteobacteria only; this can easily be changed
  mutate(mag = reorder(mag, Order, sort)) # sort by their taxonomic Order so everything shows up together

ggplot(rpkm_dat, aes(x=Sample, y=mag, col=Order)) +
  geom_point(aes(size=g_rpkm)) +
  theme(panel.background = element_blank(),
        panel.grid.major = element_line(colour = "#bdbdbd", linetype = "dotted"),
        panel.grid.minor = element_blank(),
        axis.title.y = element_text(angle = 0),
        axis.text.x = element_text(angle = 45,
                                   hjust = 1))
```
